test/compilationTests/gnosis/Utils/Math.sol

```diff
@@ -94,12 +94,18 @@ library Math {
         zpow = zpow * z / ONE;
         result += 0x9c7 * zpow / ONE;
         if (shift >= 0) {
-            if (result >> (256-shift) > 0)
+            uint tempres;
+            if (shift > 256)
+                tempres = result << uint(shift - 256);
+            else
+                tempres = result >> uint(256 - shift);
+
+            if (tempres > 0)
                 return (2**256-1);
-            return result << shift;
+            return result << uint(shift);
         }
         else
-            return result >> (-shift);
+            return result >> uint(-shift);
```

Comment:
This looks fine now. 

But I still haven't figured out how the whole code works and if the value of `shift` can be `>= 512`.